fdbserver/workloads/EncryptionOps.actor.cpp

```diff
@@ -177,19 +173,25 @@ struct EncryptionOpsWorkload : TestWorkload {
 	std::string description() const override { return "EncryptionOps"; }
 
 	Future<Void> start(Database const& cx) override {
-		for (int i = 0; i < numIterations; i++) {
-			StreamCipherKey key(AES_256_KEY_LENGTH);
-			// derive the encryption key
-			updateEncryptionKey(&key);
+		BlobCipherDomainId domainId{ 1 };
+		for (int i = 0; i < numIterations; i++, domainId++) {
+			// Step-1: Encryption key derivation, caching the cipher for later use
+			auto start = std::chrono::high_resolution_clock::now();
+			auto& cipherKeyCache = BlobCipherKeyCache::getInstance();
+			cipherKeyCache.insertCipherKey(domainId, domainId, parentCipher.get(), AES_256_KEY_LENGTH);
+			Reference<BlobCipherKey> cipherKey = cipherKeyCache.getLatestCipherKey(domainId, domainId);
+			auto end = std::chrono::high_resolution_clock::now();
+			metrics->updateKeyDerivationTime(std::chrono::duration<double, std::nano>(end - start).count());
 
 			int dataLen = isFixedSizePayload() ? pageSize : deterministicRandom()->randomInt(100, maxBufSize);
 			generateRandomData(buff.get(), dataLen);
 
-			// encrypt the payload
-			const auto& encrypted = doEncryption(&key, buff.get(), dataLen);
+			// Encrypt the payload - generates BlobCipherEncryptHeader to assist decryption later
+			BlobCipherEncryptHeader header;
+			const auto& encrypted = doEncryption(cipherKey, buff.get(), dataLen, &header);
 
-			// decrypt the payload
-			doDecryption(&key, encrypted, dataLen, buff.get(), validationBuff.get());
+			// Decrypt the payload - parses the BlobCipherEncryptHeader, fetch corresponding cipherKey and decrypt
+			doDecryption(encrypted, dataLen, header, buff.get(), validationBuff.get());
```

Comment:
Should we also run this in a non-performance test mode where we validate that the decrypted data is the same as the input, for correctness?

Reply:
doEncryption does the following:
1. Encrypt the input buffer
2. Validate encrypted buffer != original payload
3. Return encryption header information

doDecryption does the following:
1. Decrypt the buffer
2. Matches the decrypted contents == original payload

Given the time measurements are carefully placed around `EncryptEncryptBlobCipherAes265Ctr` & `DecryptBlobCipherAes256Ctr` operations, the test serves both as mini benchmark for encryption operations as well as perform correctness validation. 
Please correct me if am missing anything.